presto-verifier/src/main/java/com/facebook/presto/verifier/framework/QueryException.java

```diff
@@ -14,111 +14,70 @@
 package com.facebook.presto.verifier.framework;
 
 import com.facebook.presto.jdbc.QueryStats;
-import com.facebook.presto.spi.ErrorCode;
-import com.facebook.presto.spi.ErrorCodeSupplier;
 import com.facebook.presto.verifier.event.QueryFailure;
-import com.google.common.base.Function;
 
 import java.util.Optional;
 
-import static com.facebook.presto.verifier.framework.QueryException.Type.CLUSTER_CONNECTION;
-import static com.facebook.presto.verifier.framework.QueryException.Type.PRESTO;
+import static com.google.common.base.Preconditions.checkState;
 import static com.google.common.base.Throwables.getStackTraceAsString;
-import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
 
-public class QueryException
+public abstract class QueryException
         extends RuntimeException
 {
-    public enum Type
-    {
-        CLUSTER_CONNECTION(qe -> {
-            requireNonNull(qe, "queryException is null");
-            requireNonNull(qe.getCause(), "cause is null");
-            return qe.getCause().getClass().getSimpleName();
-        }),
-        PRESTO(qe -> {
-            requireNonNull(qe, "queryException is null");
-            return qe.prestoErrorCode.map(ErrorCodeSupplier::toErrorCode).map(ErrorCode::getName).orElse("UNKNOWN");
-        });
-
-        private final Function<QueryException, String> descriptionGenerator;
-
-        Type(Function<QueryException, String> descriptionGenerator)
-        {
-            this.descriptionGenerator = requireNonNull(descriptionGenerator, "descriptionGenerator is null");
-        }
-    }
-
-    private final Type type;
-    private final Optional<ErrorCodeSupplier> prestoErrorCode;
     private final boolean retryable;
-    private final Optional<QueryStats> queryStats;
     private final QueryStage queryStage;
 
-    private QueryException(
-            Throwable cause,
-            Type type,
-            Optional<ErrorCodeSupplier> prestoErrorCode,
-            boolean retryable,
-            Optional<QueryStats> queryStats,
-            QueryStage queryStage)
+    public QueryException(Throwable cause, boolean retryable, QueryStage queryStage)
     {
         super(cause);
-        this.type = requireNonNull(type, "type is null");
-        this.prestoErrorCode = requireNonNull(prestoErrorCode, "errorCode is null");
         this.retryable = retryable;
-        this.queryStats = requireNonNull(queryStats, "queryStats is null");
         this.queryStage = requireNonNull(queryStage, "queryStage is null");
     }
 
-    public static QueryException forClusterConnection(Throwable cause, QueryStage queryStage)
+    public abstract String getErrorCodeName();
+
+    public boolean isClusterConnectionException()
```

Comment:
@caithagoras I find it strange to include these type of methods in the abstract class. This make this abstract class tightly couples with its subclasses. It think it would be clearer to remove these methods and move the logic into the callers or move them into a separate utils class.